### PR TITLE
Add global_step in nn.py

### LINKED ISSUE #7784: Need a global counter for training.
# Problem

A global step counter to record how many times a program has executed is needed. It is needed not only by optimizers with learning rate schedule but also needed for printing metrics or debugging(such as `log_every_n` layer). 

# Solution

We can add a layer function, `get_or_create_global_step_counter()` in `layers.control_flow` module. It will return an int64 variable in CPU with a increasement operator. If the counter is created before, it will not be created again.

### PR DESCRIPTION
**Pull request by reyoung**
Fix #7784

### CODE REVIEW

File: python/paddle/fluid/learning_rate_decay.py [hunk 1]
```diff
@@ -65,32 +66,25 @@ def exponential_decay(learning_rate,
...
-    ```python
-    if not staircase:
-        decayed_learning_rate = learning_rate * exp(- decay_rate * (global_step / decay_steps))
-    else:
-        decayed_learning_rate = learning_rate * exp(- decay_rate * (global_step / decay_steps))
-    ```
+    >>> if not staircase:
+    >>>     decayed_learning_rate = learning_rate * exp(- decay_rate * (global_step / decay_steps))
+    >>> else:
+    >>>     decayed_learning_rate = learning_rate * exp(- decay_rate * (global_step / decay_steps))
```
Comment (reyoung) [inline python/paddle/fluid/learning_rate_decay.py]: This is the correct format for code list in comments.

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -3250,3 +3251,25 @@ def one_hot(input, depth):
...
+    Returns(Variable): The global run counter.
+    """
+    helper = LayerHelper('global_step_counter')
+    counter_name = '@STEP_COUNTER@'
+    counter, is_new_var = helper.create_or_get_global_variable(
+        name=counter_name, dtype='int64', shape=[1], persistable=True)
+    if is_new_var:
+        helper.set_variable_initializer(
+            counter, initializer=Constant(
+                value=0, force_cpu=True))
```
Comment (jacquesqiao) [inline python/paddle/fluid/layers/nn.py]: maybe we can start from `-1`, then we do not need to get global_step with
```cpp
global_step = global_step_counter() - 1
```

Comment (panyx0718) [reply python/paddle/fluid/layers/nn.py]: Maybe we should start from 0 and change the way global step is used.

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -3250,3 +3251,25 @@ def one_hot(input, depth):
...
+        helper.main_program.global_block().prepend_op(
+            type='increment',
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: Maybe it's better to get a global counter and explicitly increase it?

global_step = get_or_create_global_step()
global_step += 1

global step needs to be a mutable thing, maybe a non-trainable parameter.

File: python/paddle/fluid/learning_rate_decay.py [hunk 4]
```diff
@@ -30,11 +30,14 @@
...
 
 
-def exponential_decay(learning_rate,
-                      global_step,
-                      decay_steps,
-                      decay_rate,
-                      staircase=False):
+def float_global_step():
+    # the first global step is zero in learning rate decay
+    global_step = layers.global_step_counter() - 1
```
Comment (panyx0718) [inline python/paddle/fluid/learning_rate_decay.py]: this line is confusing.

File: python/paddle/fluid/learning_rate_decay.py [hunk 5]
```diff
@@ -44,16 +47,14 @@ def exponential_decay(learning_rate,
...
         decay_steps: A Python `int32` number.
         decay_rate: A Python `float` number.
         staircase: Boolean. If set true, decay the learning rate every decay_steps.
 
     Returns:
         The decayed learning rate
     """
-    if not isinstance(global_step, Variable):
-        raise ValueError("global_step is required for exponential_decay.")
+    global_step = float_global_step()
```
Comment (panyx0718) [inline python/paddle/fluid/learning_rate_decay.py]: this line feels like getting a variable, but it actually increases it?

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (reyoung) [reply python/paddle/fluid/layers/nn.py]: 1. the global step should not be a parameter. A parameter of a model should affect the output of the model. For example, `y=f(W; x)`, W is a parameter iif it will change the model behaviour.

2. The global step is a variable will be shared by `log_every_n`, `learning_rate_decays`, etc. Instead of creating several variables, here we just create a special global variable to record how many times the program has run.

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (reyoung) [reply python/paddle/fluid/layers/nn.py]: Done.

File: python/paddle/fluid/learning_rate_decay.py (hunk 4 above)
Comment (reyoung) [reply python/paddle/fluid/learning_rate_decay.py]: Done. Changed by layers.autoincrement_counter(begin=0)
